File: apple/MiniXiangqi/Settings/SettingsScreen.swift
// The Settings destination: the preferences the app keeps, and nothing else.
//
// docs/product.md, "Product navigation": Settings is the third primary
// destination, it holds the persistent preferences, it stores no game data, and
// changing one never alters an active game. It holds no interface-language
// control either — the operating system owns the language, and a control of ours
// would be a second source of truth for it.
//
// The shape is a grouped Form, ordered by what a preference reaches, nearest
// first: what the board shows (棋盘), what a move gives back (the two feedback
// switches), how an action commits (落子前确认, 删除前确认), what the next game
// opens with (人机对弈默认设置), and last the way to what the application is
// (关于). The board's two choices sit together because both are about what the
// board shows and they are independent of each other — a learner may want 图标
// discs beside the 中文 list they are learning to read — and their footer says
// how far the pair reaches, jieqi's own move reading being the one surprise
// worth a sentence. The two feedback switches are their own group with no
// header, of equal standing and neither nested under the other. 落子前确认 and
// 删除前确认 are each a group of their own so that each footer is unmistakably
// about the switch above it: one changes how a stone is placed, the other
// makes a deletion immediate, and a deletion cannot be undone. 人机对弈默认设置
// is headed and footed because its two values need saying what they are for:
// they initialize the next game's setup and never reach the game already on
// the board. 关于 is not a preference at all: it is a page rather than a
// control, so it takes a group of its own at the foot where nothing above it
// has to make room. A footer stands only where reach or consequence needs
// saying — a footer under every group is a screen nobody reads.
//
// **Settings is silent.** Sound is an event of the board, per
// docs/interaction-design.md § Sound and haptics, and a screen that clicked back
// at every switch would be the app talking about itself.
//
// Every control reads and writes through `Preferences` — the same reader its
// consumers use — rather than through `@AppStorage`, which would state each
// preference's default a second time, in a property wrapper, where a screen that
// disagreed with the board about what an unset key means would be invisible until
// somebody's first launch.

import SwiftUI

struct SettingsScreen: View {

    /// The five values, mirrored into view state and written through as they
    /// change.
    ///
    /// Mirrored because a control needs somewhere to put the value it is given,
    /// and read at construction rather than kept anywhere longer-lived because
    /// the navigation container rebuilds this screen on every visit: the mirror
    /// is never older than the visit that made it, and nothing else in the app
    /// writes these keys.
    @State private var sound = Preferences.sound.value()
    @State private var haptics = Preferences.haptics.value()
    @State private var confirmsDeletion = Preferences.deleteConfirmation.value()
    @State private var confirmsPlacement = Preferences.placementConfirmation.value()
    @State private var symbols = Preferences.pieceSymbols.value()
    @State private var notation = Preferences.notationStyle.value()
    @State private var firstMover = Preferences.defaultFirstMover.value()
    @State private var aiLevel = Preferences.defaultAiLevel.value()

    var body: some View {
        NavigationStack {
            Form {
                board
                feedback
                placement
                deletion
                humanVersusAIDefaults
                about
            }
            // The macOS-native presentation of a preference list, and the one
            // that gives a section its header and its footer.
            .formStyle(.grouped)
            .navigationTitle("nav.settings")
        }
        // Written where it is read from, as it changes: a preference screen has
        // no Save button, and the next landing or the next deletion is entitled
        // to the new answer rather than to the one the app launched with.
        .onChange(of: sound) { Preferences.sound.set(sound) }
        .onChange(of: haptics) { Preferences.haptics.set(haptics) }
        .onChange(of: confirmsDeletion) {
            Preferences.deleteConfirmation.set(confirmsDeletion)
        }
        .onChange(of: confirmsPlacement) {
            Preferences.placementConfirmation.set(confirmsPlacement)
        }
        .onChange(of: symbols) { Preferences.pieceSymbols.set(symbols) }
        .onChange(of: notation) { Preferences.notationStyle.set(notation) }
        .onChange(of: firstMover) { Preferences.defaultFirstMover.set(firstMover) }
        .onChange(of: aiLevel) { Preferences.defaultAiLevel.set(aiLevel) }
    }

    // The groups, each its own property: one Form body carrying all of them
    // stopped type-checking in reasonable time, and a section is a
    // self-contained thing anyway.

    /// The tags are the stored names themselves. A picker row is a choice being
    /// written to a preference, and the preference holds a name: a type in
    /// between would be one this screen invented for a choice whose consumer
    /// already has its own.
    private var board: some View {
        Section {
            Picker("settings.symbols.label", selection: $symbols) {
                Text("settings.symbols.hanzi").tag("hanzi")
                Text("settings.symbols.icons").tag("icons")
            }
            .accessibilityIdentifier("settings-symbols")

            Picker("settings.notation.label", selection: $notation) {
                Text("settings.notation.traditional").tag("traditional")
                Text("settings.notation.wxf").tag("wxf")
            }
            .accessibilityIdentifier("settings-notation")
        } header: {
            Text("settings.section.board")
        } footer: {
            // The scope note: the stone games have no symbols to swap, and the
            // one genuine surprise — flipping the notation preference in a
            // jieqi game and seeing its move list unmoved — is what the
            // sentence answers.
            Text("settings.board.footer")
                .accessibilityIdentifier("settings-board-footer")
        }
    }

    /// The two values a new human-versus-AI game's setup page opens with. They
    /// initialize a draft and nothing more: an active game's own side and level
    /// were frozen when it was created, and changing either of these never
    /// reaches it — which is what the footer says out loud.
    private var humanVersusAIDefaults: some View {
        Section {
            Picker("settings.defaults.firstMover", selection: $firstMover) {
                Text("setup.iMoveFirst").tag("human-first")
                Text("setup.aiMovesFirst").tag("ai-first")
                Text("setup.random").tag("random")
            }
            .accessibilityIdentifier("settings-first-mover")

            Picker("settings.defaults.aiLevel", selection: $aiLevel) {
                Text("setup.level.fast").tag("fast")
                Text("setup.level.standard").tag("standard")
                Text("setup.level.deep").tag("deep")
            }
            .accessibilityIdentifier("settings-ai-level")
        } header: {
            Text("settings.defaults.group")
        } footer: {
            Text("settings.defaults.footer")
                .accessibilityIdentifier("settings-defaults-footer")
        }
    }

    /// The two switches, of equal standing and neither conditioned on the other
    /// — except by the hardware, which conditions one of them out of existence.
    ///
    /// 触感 is **absent** on a device with no engine to drive, rather than
    /// present and dead: "on a device without them the toggle is unavailable
    /// rather than silently ineffective". `Haptics` owns the question and the
    /// reasoning for answering it by removal; what matters here is that the group
    /// survives losing a row — it has no header to be left stranded, and 声音
    /// stands alone without meaning anything different.
    private var feedback: some View {
        Section {
            Toggle("settings.sound.label", isOn: $sound)
                .accessibilityIdentifier("settings-sound")
            if Haptics.isOffered {
                Toggle("settings.haptics.label", isOn: $haptics)
                    .accessibilityIdentifier("settings-haptics")
            }
        }
    }

    /// 落子前确认: the optional pending stone of the placement games.
    ///
    /// A group of its own with a footer, the shape 删除前确认 already set for an
    /// optional confirmation — and the footer says what the flow *is* rather
    /// than warning about it, because nothing here is irreversible: what the
    /// switch buys is a second look before a stone goes down, and Undo is the
    /// recovery whichever way it stands.
    private var placement: some View {
        Section {
            Toggle("settings.confirmPlacement.label", isOn: $confirmsPlacement)
                .accessibilityIdentifier("settings-confirm-placement")
        } footer: {
            Text("settings.confirmPlacement.footer")
                .accessibilityIdentifier("settings-confirm-placement-footer")
        }
    }

    private var deletion: some View {
        Section {
            Toggle("settings.confirmDelete.label", isOn: $confirmsDeletion)
                .accessibilityIdentifier("settings-confirm-delete")
        } footer: {
            Text("settings.confirmDelete.footer")
                .accessibilityIdentifier("settings-confirm-delete-footer")
        }
    }

    /// The way to `AboutScreen`, and the screen's only row that writes nothing.
    /// It is pushed onto Settings' own stack rather than presented, because it
    /// is a page walked into from a destination's root — the platform's back
    /// control is what leaves it, and nothing about it is modal.
    private var about: some View {
        Section {
            NavigationLink("about.title") { AboutScreen() }
                .accessibilityIdentifier("settings-about")
        }
    }
}

File: windows/MiniXiangqi.Play/Settings/SettingsScreen.cs
// The Settings destination: the preferences the app keeps, and nothing else.
//
// docs/product.md, "Product navigation": Settings is the third primary
// destination, it holds the persistent preferences, it stores no game data, and
// changing one never alters an active game. It holds no interface-language
// control either — the operating system owns the language, and on Windows that
// is the system's language preference list, which .NET already resolves
// `CurrentUICulture` from.
//
// **Three groups here where the Mac has four**, and every difference is issue
// #80's owner-decided Windows trim rather than a design change:
//
//   * 棋盘 is absent entirely. Its two rows are 棋子符号 and 记谱法. The move
//     record on this platform is the core's own canonical coordinate text in
//     both languages — the 记谱法 preference and the two proper renderings
//     arrive together post-MVP — so there is nothing for the notation row to
//     choose between; and only the 汉字 symbol set is drawn, so the symbols row
//     is a preference with one option, which docs/product.md § Product
//     navigation has already ruled is not a preference. A group whose every row
//     went is a group that goes.
//   * 触感 is absent from the feedback group, **not for the MVP** rather than
//     for want of an API — Windows' own touchpad haptics exist and
//     `Motion/Feedback.cs` names them, and they are experimental, 24H2-gated and
//     rarely carried, so a switch over them now would do nothing on nearly every
//     machine. That is the contract's "unavailable rather than silently
//     ineffective" all the same, so the row is **removed rather than greyed
//     out**, which is iOS's own answer. The group survives losing it exactly as
//     the contract says it does — it has no header to strand, and the two
//     switches were never conditioned on each other, so 声音 standing alone
//     means what it meant standing above 触感.
//   * The two remaining groups are unchanged: 人机对弈默认设置, headed and
//     footed, and 删除前确认, footed.
//
// **The two-footer rule survives intact, and that is not a coincidence.** "There
// are two footers on the screen, and only two" — `settings.defaults.footer` and
// `settings.confirmDelete.footer` — and both of the groups they belong to are
// among the three that ship. Nothing here needed a third footer to explain an
// absence, because an absent row explains itself by not being there.
//
// **This screen is silent.** Sound is an event of the board, per
// docs/interaction-design.md § Sound and haptics, and a screen that clicked back
// at every switch would be the app talking about itself. There is no `Feedback`
// on this type at all, which is how that is enforced rather than remembered.
//
// **Every value is read from the store, every time it is asked for.** Nothing
// below caches: the properties are the stored preferences, so a control shows
// what *is* stored rather than what was last asked for. Two things fall out of
// that and both are wanted. A write the file system refused leaves the control
// where it was, which is the truth and needs no alert; and a preference changed
// underneath the screen — by hand, or by an Apple frontend over a shared profile
// — is picked up on the next refresh without a relaunch, which is the same
// "read at the moment of use" rule the sound gate and the deletion gate run
// under.
//
// **This is a deliberate divergence from the Mac**, which mirrors each value
// into `@State` and therefore shows what was *asked for*. That difference is
// confined to a failure mode the two platforms do not share equally: a
// `UserDefaults` write barely has one, while a file write has disk, quota and
// permissions behind it. `Changed` is raised whether or not the write landed,
// and that unconditional raise is the whole of the snap-back — it is what
// redraws the control to the stored value. Announcing only on success would
// leave a switch showing something nobody stored, silently, which is why the
// harness drives a refused write rather than describing one.

namespace MiniXiangqi.Play;

/// <summary>
/// The Settings destination, minus the window. The window turns this into XAML
/// and turns XAML's events back into calls on it, and holds nothing else — the
/// same split the Play and History destinations have, for the same reason: a
/// WinUI 3 process cannot be launched over SSH, so a screen that lived in the
/// window could be exercised on no machine this project owns.
/// </summary>
public sealed class SettingsScreen
{
    private readonly IWritablePreferences _preferences;

    public SettingsScreen(IWritablePreferences preferences) => _preferences = preferences;

    /// <summary>Raised whenever a preference was written.</summary>
    public event Action? Changed;

    /// <summary>
    /// **默认先后手** — which side a new human-versus-AI game opens on. It
    /// initializes the next pre-start page's draft and reaches no game that
    /// exists, which is what the group's footer says out loud.
    /// </summary>
    public FirstMoverChoice DefaultFirstMover => Preferences.DefaultFirstMover(_preferences);

    /// <summary>**默认 AI 等级** — how long that game's opponent thinks.</summary>
    public AiLevel DefaultAiLevel => Preferences.DefaultAiLevel(_preferences);

    /// <summary>**声音** — whether the board is heard.</summary>
    public bool Sound => Preferences.Sound(_preferences);

    /// <summary>**删除前确认** — whether deleting a History record asks first.</summary>
    public bool ConfirmsDeletion => Preferences.ConfirmsDeletion(_preferences);

    public void ChooseDefaultFirstMover(FirstMoverChoice choice)
    {
        if (choice != DefaultFirstMover)
        {
            Write(Preferences.DefaultFirstMoverKey, choice.Name());
        }
    }

    public void ChooseDefaultAiLevel(AiLevel level)
    {
        if (level != DefaultAiLevel)
        {
            Write(Preferences.DefaultAiLevelKey, level.Name());
        }
    }

    public void SetSound(bool on)
    {
        if (on != Sound)
        {
            Write(Preferences.SoundKey, on);
        }
    }

    public void SetDeleteConfirmation(bool on)
    {
        if (on != ConfirmsDeletion)
        {
            Write(Preferences.DeleteConfirmationKey, on);
        }
    }

    /// <summary>
    /// The written form is the *name*, never an index: an index is a position in
    /// a list a later design may reorder and a name is not, and these three names
    /// are docs/game-data.md's serialized vocabulary — the preference, the frozen
    /// configuration and the archive all say the same word.
    ///
    /// A preference already holding the chosen value is not written again, above.
    /// That is not an optimisation over a cheap write: a control that reports the
    /// value it was just *given* is what a XAML selector does when a refresh sets
    /// it, and a screen that wrote on every such report would rewrite the file
    /// every time it drew.
    ///
    /// **`Changed` is raised whether or not the write landed, and that is
    /// load-bearing rather than lax.** A refused write is answered by the control
    /// snapping back to the stored value, and this raise is the only thing that
    /// redraws it — the store swallowed the failure and has nothing to report.
    /// Announcing only on success would leave a switch showing a value nobody
    /// stored, and nothing else in the app would notice.
    /// </summary>
    private void Write(string key, string value)
    {
        _preferences.Write(key, value);
        Changed?.Invoke();
    }

    private void Write(string key, bool value)
    {
        _preferences.Write(key, value);
        Changed?.Invoke();
    }
}
